rasa/core/actions/action.py

```diff
@@ -940,3 +949,141 @@ def name(self) -> Text:
+    async def run(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+    ) -> List[Event]:
+        """Runs action. Please see parent class for the full docstring."""
+        slot_events: List[Event] = []
+
+        default_slots = {
+            rasa.shared.core.constants.REQUESTED_SLOT,
+            rasa.shared.core.constants.SESSION_START_METADATA_SLOT,
+            rasa.shared.core.constants.SLOT_LISTED_ITEMS,
+            rasa.shared.core.constants.SLOT_LAST_OBJECT,
+            rasa.shared.core.constants.SLOT_LAST_OBJECT_TYPE,
+        }
+        user_slots = [slot for slot in domain.slots if slot.name not in default_slots]
+
+        for slot in user_slots:
+            for mapping in slot.mappings:
+                # skip to the next mapping because a mapping with conditions
+                # is applicable only within the context of an active loop
+                if mapping.get("conditions"):
+                    continue
+
+                intent_is_desired = SlotMapping.intent_is_desired(
+                    mapping, tracker, domain
+                )
+
+                should_fill_trigger_slot = (
+                    mapping["type"] == str(SlotMapping.FROM_TRIGGER_INTENT)
+                    and intent_is_desired
+                )
+
+                if should_fill_trigger_slot:
+                    # skip, only applicable in an active loop
+                    continue
+
+                should_fill_custom_slot = (
+                    mapping["type"] == str(SlotMapping.CUSTOM) and intent_is_desired
+                )
+
+                should_fill_entity_slot = (
+                    mapping["type"] == str(SlotMapping.FROM_ENTITY)
+                    and intent_is_desired
+                    and SlotMapping.entity_is_desired(mapping, tracker,)
+                )
+
+                should_fill_intent_slot = (
+                    mapping["type"] == str(SlotMapping.FROM_INTENT)
+                    and intent_is_desired
+                )
+
+                should_fill_text_slot = (
+                    mapping["type"] == str(SlotMapping.FROM_TEXT) and intent_is_desired
+                )
+
+                value: List[Any] = []
+                if should_fill_entity_slot:
+                    value = list(
+                        tracker.get_latest_entity_values(
+                            mapping.get(ENTITY_ATTRIBUTE_TYPE),
+                            mapping.get(ENTITY_ATTRIBUTE_ROLE),
+                            mapping.get(ENTITY_ATTRIBUTE_GROUP),
+                        )
+                    )
+                elif should_fill_intent_slot:
+                    value = [mapping.get("value")]
+                elif should_fill_text_slot:
+                    value = [tracker.latest_message.text]
+                elif should_fill_custom_slot:
+                    custom_action = mapping.get("action")
```

Comment:
I would consider moving the execution of a custom action to a separate function — just because it can be treated as an individual unit and because the `run` function is quite huge already.